### PR TITLE
blockchain & p2p: transfer coding of custom messages to readerwriter crate

### PR DESCRIPTION
**Pull request by p0lunin**
This PR is a part #424, which implement encoding/decoding of [blockchain::protocol::Message](https://github.com/stellar/slingshot/blob/main/blockchain/src/protocol.rs#L537).

### CODE REVIEW

File: blockchain/src/codec.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{utreexo, BlockHeader, BlockID, BlockTx};
+    use curve25519_dalek::ristretto::CompressedRistretto;
+    use curve25519_dalek::scalar::Scalar;
+    use p2p::reexport::BytesMut;
```
Comment (oleganza) [inline blockchain/src/codec.rs:327]: I think we can use `Vec<u8>` to test encoding, no need to pull big dependency (even if you'd make it a dev-dependency).

File: blockchain/Cargo.toml [hunk 2]
```diff
@@ -29,6 +29,12 @@ path = "../zkvm"
 [dependencies.starsig]
 path = "../starsig"
 
+[dependencies.p2p]
+path = "../p2p"
```
Comment (oleganza) [inline blockchain/Cargo.toml:33]: if we only need this for the test, it should be in dev-dependencies. But i think we can just use Vec<u8> and not pull p2p/tokio/bytes.

File: blockchain/src/codec.rs (hunk 1 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: p2p was removed from dependencies in blockchain

File: blockchain/Cargo.toml (hunk 2 above)
Comment (p0lunin) [reply blockchain/Cargo.toml]: p2p was removed from dependencies in blockchain

File: zkvm/src/constraints.rs [hunk 3]
```diff
@@ -12,6 +12,7 @@ use subtle::{ConditionallySelectable, ConstantTimeEq};
 use crate::encoding::*;
 use crate::errors::VMError;
 use crate::scalar_witness::ScalarWitness;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/constraints.rs:15]: Since we do `use crate::encoding::*` in order to get all the traits we need, why not put `pub use readerwriter::Encodable` in that module, so in all other places we don't have to add `use readerwriter::Encodable`? See also another suggestion in `src/encoding.rs`.

```suggestion
```

File: zkvm/src/encoding.rs [hunk 4]
```diff
@@ -7,22 +7,6 @@ pub use readerwriter::{ReadError, Reader, WriteError, Writer};
 
 use crate::errors::VMError;
```
Comment (oleganza) [inline zkvm/src/encoding.rs:9]: Per suggestion above - let's pub-export Encodable here, so we can continue `use crate::encoding::*`
```suggestion
pub use readerwriter::Encodable;
```

File: zkvm/src/verifier.rs [hunk 5]
```diff
@@ -4,10 +4,11 @@ use bulletproofs::{BulletproofGens, PedersenGens};
 use curve25519_dalek::ristretto::CompressedRistretto;
 use merlin::Transcript;
 use musig::{Multisignature, VerificationKey};
+use readerwriter::Encodable;
 
 use crate::constraints::Commitment;
 use crate::contract::ContractID;
-use crate::encoding::{Encodable, Reader};
+use crate::encoding::Reader;
```
Comment (oleganza) [inline zkvm/src/verifier.rs:11]: per suggestions above:

```suggestion
use crate::encoding::{Encodable, Reader};
```

File: zkvm/src/verifier.rs [hunk 6]
```diff
@@ -4,10 +4,11 @@ use bulletproofs::{BulletproofGens, PedersenGens};
 use curve25519_dalek::ristretto::CompressedRistretto;
 use merlin::Transcript;
 use musig::{Multisignature, VerificationKey};
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/verifier.rs:7]:
```suggestion
```

File: zkvm/src/types.rs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 use curve25519_dalek::scalar::Scalar;
 use merlin::Transcript;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/types.rs:5]: per suggestions above:

```suggestion
```

File: zkvm/src/tx.rs [hunk 8]
```diff
@@ -12,6 +12,7 @@ use crate::fees::FeeRate;
 use crate::merkle::{Hash, MerkleItem, MerkleTree};
 use crate::transcript::TranscriptProtocol;
 use crate::verifier::Verifier;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/tx.rs:15]: per suggestions above:

```suggestion
```

File: zkvm/src/scalar_witness.rs [hunk 9]
```diff
@@ -6,6 +6,7 @@ use spacesuit::SignedInteger;
 
 use crate::encoding::*;
 use crate::errors::VMError;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/scalar_witness.rs:9]: per suggestions above:

```suggestion
```

File: zkvm/src/prover.rs [hunk 10]
```diff
@@ -4,11 +4,11 @@ use bulletproofs::{BulletproofGens, PedersenGens};
 use curve25519_dalek::ristretto::CompressedRistretto;
 use merlin::Transcript;
 use musig::VerificationKey;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/prover.rs:7]: per suggestions above:

```suggestion
```

File: zkvm/src/program.rs [hunk 11]
```diff
@@ -8,6 +8,7 @@ use crate::types::String;
 
 use core::borrow::Borrow;
 use merlin::Transcript;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/program.rs:11]: per suggestions above:

```suggestion
```

File: zkvm/src/predicate.rs [hunk 12]
```diff
@@ -17,6 +17,7 @@ use crate::errors::VMError;
 use crate::merkle::{Hash, Hasher, MerkleItem, MerkleTree, Path};
 use crate::program::{Program, ProgramItem};
 use crate::transcript::TranscriptProtocol;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/predicate.rs:20]: per suggestions above:

```suggestion
```

File: zkvm/src/ops.rs [hunk 13]
```diff
@@ -8,6 +8,7 @@ use crate::errors::VMError;
 use crate::program::ProgramItem;
 use crate::scalar_witness::ScalarWitness;
 use crate::types::String;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/ops.rs:11]: per suggestions above:

```suggestion
```

File: zkvm/src/merkle.rs [hunk 14]
```diff
@@ -3,6 +3,7 @@ use crate::encoding::*;
 use crate::errors::VMError;
 use core::marker::PhantomData;
 use merlin::Transcript;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/merkle.rs:6]: per suggestions above:

```suggestion
```

File: zkvm/src/contract.rs [hunk 15]
```diff
@@ -9,6 +9,7 @@ use crate::program::ProgramItem;
 use crate::transcript::TranscriptProtocol;
 use crate::types::{String, Value};
 use merlin::Transcript;
+use readerwriter::Encodable;
```
Comment (oleganza) [inline zkvm/src/contract.rs:12]: per suggestions above:

```suggestion
```

File: blockchain/src/codec.rs [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    fn read_blockid(&mut self) -> Result<BlockID, ReadError> {
+        self.read_u8x32().map(BlockID)
+    }
+
+    fn read_hash(&mut self) -> Result<Hash, ReadError> {
+        self.read_u8x32().map(Hash)
+    }
+
+    fn read_block_header(&mut self) -> Result<BlockHeader, ReadError> {
+        Ok(BlockHeader {
```
Comment (oleganza) [inline blockchain/src/codec.rs:81]: Maybe `impl Decodable` for `blockchain::BlockHeader`?

File: blockchain/src/codec.rs [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+            tip_signature: self.read_signature()?,
+            shortid_nonce: self.read_u64()?,
+            shortid_list: self.read_shortid_vec()?,
+        })
+    }
+
+    fn read_tx(&mut self) -> Result<Tx, ReadError> {
+        let tx_size = self.read_u32()? as usize;
+        let vec = self.read_vec(tx_size)?;
+        Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
```
Comment (oleganza) [inline blockchain/src/codec.rs:105]: Maybe `impl Decodable` for `zkvm::Tx`?

File: blockchain/src/codec.rs [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+    }
+
+    fn read_txs(&mut self) -> Result<Vec<BlockTx>, ReadError> {
+        let len = self.read_u32()? as usize;
+        const BLOCK_TX_MIN_LENGTH: usize = 8 + 8 + 8 + 8 + 32 + 32 + 1 + 11 * 32 + 8;
+        self.read_vec_with(len, BLOCK_TX_MIN_LENGTH, |src| {
+            let tx = src.read_tx()?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:112]: Maybe `impl Decodable` for `blockchain::BlockTx`?

File: blockchain/src/codec.rs [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+    fn write_blockid(
+        &mut self,
+        label: &'static [u8],
+        block_id: &BlockID,
+    ) -> Result<(), WriteError> {
+        self.write(label, block_id.as_ref())
+    }
+
+    fn write_hash(&mut self, label: &'static [u8], hash: &Hash) -> Result<(), WriteError> {
+        self.write(label, hash.0.as_ref())
```
Comment (oleganza) [inline blockchain/src/codec.rs:156]: there's `impl Deref for Hash` with `Target = [u8]`, so we don't even need this helper. Just write `write(b"foo", &hash)`.

File: blockchain/src/codec.rs [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+            4 => Ok(MessageType::MempoolTxs),
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+fn read<T>(data: Result<T, ReadError>, label: &'static str) -> Result<T, ReadError> {
```
Comment (oleganza) [inline blockchain/src/codec.rs:41]: This helper name is super-confusing, as it's mapping the error, not reading anything.
Do we really need it? It maps ReadError into ReadError, maybe leave it as is? 

If you really want to re-wrap it, i'd make a func `rewrap_err(ReadError,..)->ReadError` and use it as `foo().map_err(|e| rewrap_err(e, b"foo failed"))?`

File: blockchain/src/codec.rs [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+            height: self.read_u64()?,
+            prev: self.read_blockid()?,
+            timestamp_ms: self.read_u64()?,
+            txroot: self.read_hash()?,
+            utxoroot: self.read_hash()?,
+            ext: self.read_u8_vec()?,
+        })
+    }
+
+    fn read_inventory(&mut self) -> Result<Inventory, ReadError> {
```
Comment (oleganza) [inline blockchain/src/codec.rs:92]: this can probably be a impl Decoder for Inventory, similarly for all other messages?

File: blockchain/src/codec.rs [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+    })
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
```
Comment (oleganza) [inline blockchain/src/codec.rs:59]: little convenience helpers like `read_signature` for external types, not tied to the blockchain encoding business are totally fine to have in the ReaderExt (like ZkVM uses `read_point`, `read_scalar`). But big fat methods for the types we have, like Inventory, or Block - can totally be defined as `impl Decodable` in appropriate places.

File: blockchain/src/codec.rs (hunk 19 above)
Comment (oleganza) [reply blockchain/src/codec.rs:156]: Oh, let's keep it for symmetry with `read_hash`, which is nice because it does the type wrapping.

File: zkvm/src/constraints.rs (hunk 3 above)
Comment (p0lunin) [reply zkvm/src/constraints.rs]: Ok.

File: zkvm/src/encoding.rs (hunk 4 above)
Comment (p0lunin) [reply zkvm/src/encoding.rs:9]: Ok.

File: zkvm/src/verifier.rs (hunk 5 above)
Comment (p0lunin) [reply zkvm/src/verifier.rs]: Ok.

File: zkvm/src/verifier.rs (hunk 6 above)
Comment (p0lunin) [reply zkvm/src/verifier.rs]: Ok.

File: zkvm/src/types.rs (hunk 7 above)
Comment (p0lunin) [reply zkvm/src/types.rs]: Ok.

File: zkvm/src/tx.rs (hunk 8 above)
Comment (p0lunin) [reply zkvm/src/tx.rs]: Ok.

File: zkvm/src/scalar_witness.rs (hunk 9 above)
Comment (p0lunin) [reply zkvm/src/scalar_witness.rs]: Ok.

File: zkvm/src/prover.rs (hunk 10 above)
Comment (p0lunin) [reply zkvm/src/prover.rs]: Ok.

File: zkvm/src/program.rs (hunk 11 above)
Comment (p0lunin) [reply zkvm/src/program.rs]: Ok.

File: zkvm/src/predicate.rs (hunk 12 above)
Comment (p0lunin) [reply zkvm/src/predicate.rs]: Ok.

File: zkvm/src/ops.rs (hunk 13 above)
Comment (p0lunin) [reply zkvm/src/ops.rs]: Ok.

File: zkvm/src/merkle.rs (hunk 14 above)
Comment (p0lunin) [reply zkvm/src/merkle.rs]: Ok.

File: zkvm/src/contract.rs (hunk 15 above)
Comment (p0lunin) [reply zkvm/src/contract.rs]: Ok.

File: blockchain/src/codec.rs (hunk 20 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Agree with you, this function is excess

File: blockchain/src/codec.rs (hunk 22 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:157]: Removed big fat methods. Added impl Decodable/Encodable

File: blockchain/src/codec.rs (hunk 21 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Added impl Decodable/Encodable

File: blockchain/src/codec.rs (hunk 18 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Moved these functions outside of ReaderExt, because BlockTx implement Encodable already.

File: blockchain/src/codec.rs (hunk 17 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Moved these functions outside of ReaderExt, because Tx implement Encodable already.

File: blockchain/src/codec.rs (hunk 19 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:209]: Added impl Decodable/Encodable

File: blockchain/src/codec.rs (hunk 16 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Added impl Decodable/Encodable

File: blockchain/src/codec.rs [hunk 23]
```diff
@@ -0,0 +1,388 @@
...
+        w.write_u64(b"version", self.version)?;
+        self.tip.encode(w)?;
+        w.write_signature(&self.tip_signature)?;
+        w.write_u64(b"shortid_nonce", self.shortid_nonce)?;
+        w.write_shortid_vec(b"shortid_list", &self.shortid_list)?;
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + self.tip.encoded_length() + 32 + 8 + 4 + self.shortid_list.len()
```
Comment (oleganza) [inline blockchain/src/codec.rs:83]: 1) signature is 64 bytes long.
2) shortid_list.len() is in items, not bytes. 

Note for later: ShortID should verify that the buffer is divisible by SHORTID_LEN so we don't accidentally have extra bytes that will be re-encoded, but not converted into IDs. Right now there's such possibility (see .map(ShortIDVec) below).

```suggestion
        8 + self.tip.encoded_length() + 64 + 8 + 4 + self.shortid_list.len()*SHORTID_LEN
```

File: blockchain/src/codec.rs [hunk 24]
```diff
@@ -0,0 +1,388 @@
...
+// Tx now implement Encoder, but it's implementation isn't suitable for
+// encoding in that case
+fn read_tx(buf: &mut impl Reader) -> Result<Tx, ReadError> {
+    let tx_size = buf.read_u32()? as usize;
+    let vec = buf.read_vec(tx_size)?;
+    Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
+}
+
+fn write_tx(tx: &Tx, dst: &mut impl Writer) -> Result<(), WriteError> {
+    dst.write_u32(b"tx length", tx.encoded_length() as u32)?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:108]: i'd strive for using "identifier-friendly" labels, so they can be used in something like JSON, if needed.

```suggestion
    dst.write_u32(b"tx_length", tx.encoded_length() as u32)?;
```

File: blockchain/src/codec.rs [hunk 25]
```diff
@@ -0,0 +1,388 @@
...
+}
+
+fn read_block_tx(src: &mut impl Reader) -> Result<BlockTx, ReadError> {
+    let tx = read_tx(src)?;
+    let len = src.read_u32()? as usize;
+    let proofs = src.read_vec_with(len, 1, |src| match src.read_u8()? {
+        0 => Ok(Proof::Transient),
+        1 => merkle::Path::decode(src)
+            .map(Proof::Committed)
+            .map_err(|_| ReadError::InvalidFormat),
```
Comment (oleganza) [inline blockchain/src/codec.rs:119]: doesn't `Path::decode` return ReadError already?

```suggestion
,
```

Comment (p0lunin) [reply blockchain/src/codec.rs:119]: https://github.com/p0lunin/slingshot/blob/transfer_p2p_coding_to_readwriter/zkvm/src/merkle.rs#L367 no, it's return VMError

File: blockchain/src/codec.rs [hunk 26]
```diff
@@ -0,0 +1,388 @@
...
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
+        let s = self.read_u8x32()?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:158]: it's shorter to read a 64-byte buffer and use Signature::from_bytes()

File: blockchain/src/codec.rs [hunk 27]
```diff
@@ -0,0 +1,388 @@
...
+        let r = self.read_u8x32()?;
+        Ok(Signature {
+            s: Scalar::from_bits(s),
+            R: CompressedRistretto(r),
+        })
+    }
+
+    fn read_shortid_vec(&mut self) -> Result<ShortIDVec, ReadError> {
+        self.read_u8_vec().map(ShortIDVec)
```
Comment (oleganza) [inline blockchain/src/codec.rs:167]: I would make vec private so we don't have trailing bytes and mess up encoding length calculation, and make a fallible `new(Vec<u8>)->Option<ShortIDVec>` method:

```suggestion
        self.read_u8_vec().map(ShortIDVec::new).ok_or(ReadError::InsufficientBytes).
```
```suggestion
        self.read_u8_vec().map(ShortIDVec)
```
```suggestion
        self.read_u8_vec().map(ShortIDVec)
```

(we can do it in a separate PR)

File: blockchain/src/codec.rs [hunk 28]
```diff
@@ -0,0 +1,388 @@
...
+    }
+
+    fn read_hash(&mut self) -> Result<Hash, ReadError> {
+        self.read_u8x32().map(Hash)
+    }
+}
+
+trait WriterExt: Writer + Sized {
+    fn write_u8_vec(&mut self, label: &'static [u8], vec: &[u8]) -> Result<(), WriteError> {
+        self.write_u32(b"vec length", vec.len() as u32)?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:181]: We mostly used short `"len"` label for all length prefixes.
```suggestion
        self.write_u32(b"len", vec.len() as u32)?;
```

File: blockchain/src/codec.rs [hunk 29]
```diff
@@ -0,0 +1,388 @@
...
+        self.write(label, vec)?;
+        Ok(())
+    }
+
+    fn write_signature(&mut self, sig: &Signature) -> Result<(), WriteError> {
+        self.write(b"scalar", sig.s.as_bytes())?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:187]: Use the canonical 64-byte signature encoding in a single `write(b"sig", &sig.to_bytes()[..])`.

File: blockchain/src/codec.rs [hunk 30]
```diff
@@ -0,0 +1,388 @@
...
+                dst.write_u8(b"message type", MessageType::GetMempoolTxs as u8)?;
+                dst.write_u64(b"shortid_nonce", g.shortid_nonce)?;
+                dst.write_shortid_vec(b"shortid_list", &g.shortid_list)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        unimplemented!()
```
Comment (oleganza) [inline blockchain/src/codec.rs:304]: one last step for the man, giant leap for the mankind!

File: blockchain/src/codec.rs (hunk 25 above)
Comment (oleganza) [reply blockchain/src/codec.rs:119]: oh, dang, gotta update that one two.

File: blockchain/src/codec.rs [hunk 31]
```diff
@@ -0,0 +1,388 @@
...
+
+impl Decodable for Message {
+    fn decode(src: &mut impl Reader) -> Result<Self, ReadError>
+    where
+        Self: Sized,
+    {
+        let message_type_byte = src.read_u8()?;
+        let message_type = MessageType::try_from(message_type_byte)?;
+        match message_type {
+            MessageType::Block => {
```
Comment (oleganza) [inline blockchain/src/codec.rs:224]: it's going to be easier to review these pairs of read/write methods for each type, if they were grouped under `Message::(en|de)code_$type(&mut impl Reader)`: e.g. `decode_block`, `decode_getblock` etc.

File: blockchain/src/codec.rs [hunk 32]
```diff
@@ -0,0 +1,388 @@
...
+            }
+        }
+    }
+}
+
+impl Encodable for Message {
+    fn encode(&self, dst: &mut impl Writer) -> Result<(), WriteError> {
+        match self {
+            Message::Block(b) => {
+                dst.write_u8(b"message type", MessageType::Block as u8)?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:271]: as mentioned above, maybe we use underscores `_` instead of spaces in the labels?

```suggestion
                dst.write_u8(b"message_type", MessageType::Block as u8)?;
```

File: blockchain/src/codec.rs (hunk 23 above)
Comment (oleganza) [reply blockchain/src/codec.rs:83]: Filed here: https://github.com/stellar/slingshot/issues/436

File: blockchain/src/codec.rs (hunk 24 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:108]: Ok.

File: blockchain/src/codec.rs (hunk 26 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:158]: Ok.

File: blockchain/src/codec.rs (hunk 28 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:181]: Ok.

File: blockchain/src/codec.rs (hunk 29 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:187]: Ok.

File: blockchain/src/codec.rs (hunk 32 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:271]: Ok.

File: blockchain/src/codec.rs (hunk 30 above)
Comment (p0lunin) [reply blockchain/src/codec.rs:298]: We will remove this method from Encodable trait (see #437)

File: blockchain/src/codec.rs (hunk 23 above)
Comment (p0lunin) [reply blockchain/src/codec.rs]: Done

File: blockchain/src/codec.rs [hunk 33]
```diff
@@ -0,0 +1,406 @@
...
+        Ok(Message::Block(Block {
+            header,
+            signature,
+            txs,
+        }))
+    }
+
+    fn encode_get_block(g: &GetBlock, dst: &mut impl Writer) -> Result<(), WriteError> {
+        dst.write_u8(b"message_type", MessageType::GetBlock as u8)?;
+        dst.write_u64(b"block height", g.height)?;
```
Comment (oleganza) [inline blockchain/src/codec.rs:230]:
```suggestion
        dst.write_u64(b"block_height", g.height)?;
```

File: blockchain/src/codec.rs (hunk 27 above)
Comment (oleganza) [reply blockchain/src/codec.rs:161]: Cf. https://github.com/stellar/slingshot/issues/436
